Set and restore raw mode using `x/term` on Unix systems

niten94 opened the pull request:
This pull request switches all Unix platforms to only one implementation, that uses some functions in `x/term` like upstream to set the terminal to raw mode and restore it. This is done to avoid related bugs on new platform releases by relying on a module instead, and also deduplicate code.

This was mainly done to fix a bug on OpenBSD 7.5 or newer where the screen doesn't start, which also requires builds to be done using Go 1.23 or newer.

Previous implementations also specifically had at least one of these issues:
- incorrect usage of `Syscall*()` (BSD systems)
- `VMIN` and `VTIME` not set (BSD systems)
- termios not copied as a whole prior to modification (Linux, Solaris)

There had been no issues when I tested on Linux, OpenBSD, FreeBSD and OpenIndiana (illumos distribution, like Solaris).

This pull request replaces #2 which is similar, but it is slightly modified to present changes better.

Related: zyedidia/micro#3557

niten94 committed:
Merge tscreen_windows.go with tscreen_stub.go

niten94 committed:
Set and restore raw mode using x/term on Linux

On Linux, set the terminal to raw mode and restore it using MakeRaw()
and Restore() in x/term respectively, which is used upstream in place of
x/sys/unix.IoctlSetTermios() or syscall.Syscall6() in other platforms.

MakeRaw() performs the same changes with certain termios fields, but
copies the whole struct prior instead of those fields only.

niten94 committed:
Use Linux screen implementation on most Unix systems

Switch to Linux screen implementation on all Unix platforms excluding
MacOS, and add Unix platforms that are supported upstream only to the
build constraint.

The modifications on certain termios fields only differ on BSD systems
where VMIN and VTIME were not modified.

niten94 commented:
@JoeKar @dmaluka

Only new OpenBSD releases will benefit currently if this PR is merged, but may this at least be reviewed sometime, since it has been more than 1 year since such releases where Micro stopped running?

I think I've read on OpenBSD mailing lists that Tcell cannot be patched in their package of Micro due to being a Go dependency, but I don't remember which thread I have read this.

dmaluka commented:
Quick questions (without looking into this in any depth):
1. Is this a v2.0.14 regression?
2. Are these fixes applicable to upstream tcell?

I guess we cannot guarantee a fast review, but we can add https://github.com/zyedidia/micro/issues/3557 to the list of blockers for v2.0.15 (https://github.com/zyedidia/micro/milestone/3) if needed.

niten94 commented:
> Quick questions (without looking into this in any depth):
> 
> 1. Is this a v2.0.14 regression?
> 2. Are these fixes applicable to upstream tcell?

1. No, but this fixes a bug that doesn't occur upstream. OpenBSD 7.5 and newer is currently the only platform where the screen cannot start, due to a lot of factors including the removal of [`syscall(2)`](https://man.openbsd.org/OpenBSD-7.4/syscall.2) which are addressed in Go 1.23 and `x/term`.

2. No, since upstream Tcell already switched to one generic implementation that uses `x/term`:
https://github.com/gdamore/tcell/blob/d86b440bcb24a485e77211ed87edb91c6f600fe5/tty_unix.go#L83

> I guess we cannot guarantee a fast review

I could understand that this PR may not be reviewed early or fast, due to probably having less priority than other issues that are more important to Micro itself.

Still, I believe that this PR shouldn't take too long to review since it switches to functions in `x/term` like upstream as a somewhat quick fix, which are similar to the code replaced in this PR.

> but we can add [zyedidia/micro#3557](https://github.com/zyedidia/micro/issues/3557) to the list of blockers for v2.0.15 (https://github.com/zyedidia/micro/milestone/3) if needed.

It may not seem pleasant for users on a specific platform that Micro remains broken 1 year after the previous version, so I would suggest to do this.

JoeKar commented:
I agree to declare it as a blocker.We‘ve to fix it downstream like suggested, since we‘re unfornunately far away from using upstream tcell directly.Am 03.10.2025 um 15:36 schrieb niten94 ***@***.***>:﻿niten94 left a comment (micro-editor/tcell#3)

Quick questions (without looking into this in any depth):

Is this a v2.0.14 regression?
Are these fixes applicable to upstream tcell?




No, but this fixes a bug that doesn't occur upstream. OpenBSD 7.5 and newer is currently the only platform where the screen cannot start, due to a lot of factors including the removal of syscall(2) which are addressed in Go 1.23 and x/term.


No, since upstream Tcell already switched to one generic implementation that uses x/term:
https://github.com/gdamore/tcell/blob/d86b440bcb24a485e77211ed87edb91c6f600fe5/tty_unix.go#L83



I guess we cannot guarantee a fast review

I could understand that this PR may not be reviewed early or fast, due to probably having less priority than other issues that are more important to Micro itself.
Still, I believe that this PR shouldn't take too long to review since it switches to functions in x/term like upstream as a somewhat quick fix, which are similar to the code replaced in this PR.

but we can add zyedidia/micro#3557 to the list of blockers for v2.0.15 (https://github.com/zyedidia/micro/milestone/3) if needed.

It may not seem pleasant for users on a specific platform that Micro remains broken 1 year after the previous version, so I would suggest to do this.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

dmaluka commented:
I've looked through the changes, looks quite sane. But I haven't tested it on anything other than Linux.

JoeKar commented on tscreen_unix_darwin.go:
```diff
@@ -0,0 +1,44 @@
+package tcell
```
Don't we need the `build` define here?

`// +build darwin`

Otherwise I'm a bit puzzled which interface `tscreen_unix.go` will use to apply the right TTY handling (polling or not).

niten94 replied:
No, but I have added it now to be consistent with upstream and other files downstream. As mentioned in `$ go help buildconstraint`, `_GOOS` in the filename is recognized as a build constraint.

JoeKar replied:
> As mentioned in `$ go help buildconstraint`, `_GOOS` in the filename is recognized as a build constraint.

Ah, didn't know this. Thanks!

JoeKar commented on tscreen_unix_other.go:
```diff
@@ -0,0 +1,24 @@
+package tcell
```
Like above?
`// +build aix dragonfly freebsd linux netbsd openbsd solaris zos`

niten94 replied:
Yes, I forgot to insert that line which will cause build errors on other platforms.

niten94 committed:
Use generic Unix screen implementation on MacOS

The differences with the original implementation are the same with BSD
systems.

JoeKar commented:
In `micro`'s [go.mod#L11](https://github.com/zyedidia/micro/blob/815ca0b6d840bc619d56fc9571cd6b9f8f8e233c/go.mod#L11) we can now refer to `tcell`'s [v2.0.12](https://github.com/micro-editor/tcell/releases/tag/v2.0.12) tag.

niten94 commented:
I'll make a PR on Micro this week and try to post a notice to the OpenBSD mailing list.

I have also noticed that a workflow failed on this repository, but I was able to run `go build -v .` on Tcell using Go 1.19. It seems to be caused by the workflow building using Go 1.13, and old build constraint comments being removed in the updated `x/sys` version.

The workflow didn't run on my fork due to the reason displayed as below in the Actions tab:
<img width="463" height="220" alt="image" src="https://github.com/user-attachments/assets/eb359735-b96a-4374-bf5e-5c5dbbb834d6" />

It probably doesn't need to be addressed, since the legacy branch will unlikely be updated much. I also plan to discuss and start the sync with upstream after current tasks like the release of 2.0.15 are done.

JoeKar commented:
I was also noticed by mail and fixed the legacy workflow with the min. required go version `1.18`. :wink: 
So we can use [v2.0.13](https://github.com/micro-editor/tcell/tree/v2.0.13) now.

BTW:
> The workflow didn't run on my fork due to the reason displayed as below in the Actions tab:

This can be changed in you fork with the repository settings...
<img width="907" height="409" alt="grafik" src="https://github.com/user-attachments/assets/94c62a7d-0226-4f3a-96b4-359fbdfc423a" />
